README
    p7zip 17.05
    ===========

Homepage : https://github.com/p7zip-project/p7zip

p7zip is a port of the Windows programs 7z.exe and 7za.exe provided by 7-zip.

7-zip is a file archiver with the highest compression ratio.
Homepage : www.7-zip.org

  7z uses plugins (7z.so and Codecs/Rar.so) to handle archives.
  7za is a stand-alone executable (7za handles less archive formats than 7z).
  7zr is a light stand-alone executable that supports only 7z/LZMA/BCJ/BCJ2.

  Some GUI :
    7zG  : beta  stage
    7zFM : alpha stage

CAUTION :
---------

- FIRST : DO NOT USE the 7-zip format for backup purpose on Linux/Unix because :
  - 7-zip does not store the owner/group of the file

  On Linux/Unix, in order to backup directories you must use tar !
  to backup a directory  : tar cf - directory | 7za a -si directory.tar.7z
  to restore your backup : 7za x -so directory.tar.7z | tar xf -

- if you want to send files and directories (not the owner of file)
  to others Unix/MacOS/Windows users, you can use the 7-zip format.

  example : 7za a directory.7z  directory

  do not use "-r" because this flag does not do what you think
  do not use "directory/*" because of ".*" files
   (example : "directory/*" does not match "directory/.profile")

BUILD :
-------

  If you have downloaded the "bin" package,
    use directly the program bin/7za.   (tested on Debian 3, Ubuntu 14.10)
    As the program is statically linked, its should run on many x86 or amd64 linux.

  If you have downloaded the "source" package,

    According to your OS, copy makefile.linux,
    makefile.freebsd, makefile.cygwin, ...
        over makefile.machine

  Example : building 7za, 7z (with its plugins), 7zr and Client7z and passing the internal tests
    cp makefile.linux_x86_asm_gcc_4 makefile.linux
    make all_test


    If you want to make a parallel build on a 4 cpu machine : make -j 4 TARGET

    If you have trouble, try : make -f makefile.oldmake TARGET

    make depend   : to rebuild the makefile.depend
    make clean    : to clean all directories
    make          : to build bin/7za
    make sfx      : to build bin/7zCon.sfx (7za can now create SFX archive)
    make 7z       : to build bin/7z and its plugins :
                          - "bin/7z.so" (GNU LGPL + AES code license)
                          - "bin/Codecs/Rar.so" (GNU LGPL + unRAR restriction)
    make 7zr      : to build bin/7zr
    make all      : to build bin/7za and bin/7zCon.sfx
    make all2     : to build bin/7za, bin/7z (with its plugins) and bin/7zCon.sfx
    make all3     : to build bin/7za, bin/7z (with its plugins), bin/7zr and bin/7zCon.sfx
    make test     : to test bin/7za (extracting, archiving, password ...)
    make test_7z  : to test bin/7z  (extracting, archiving, password ...)
    make test_7zr : to test bin/7zr (extracting, archiving, ...)

    make 7zG      : to build bin/7zG and its plugins :
                          - "bin/7z.so" (GNU LGPL + AES code license)
                          - "bin/Codecs/Rar.so" (GNU LGPL + unRAR restriction)
    make test_7zG : to test bin/7zG (extracting, archiving, ...)


  this procedure has been tested on :
   - x86 CPU :
    Linux - Debian 3.0 Stable
    Linux - Ubuntu 14.10
    Linux - Ubuntu 15.10    
    MacOS 10.6.6

   - AMD64 CPU :
    Linux - Ubuntu 14.10
    Linux - Ubuntu 15.10     
    MacOS 10.6.6

   - powerpc CPU :
    MacOS X 10.4 (ppc)

   - arm :
    Android (Galaxy Note 3)

  Some older versions of p7zip were tested :
   - x86 CPU :
    DOS   - (built with DJGPP, see http://blairdude.googlepages.com/p7zip )
    Linux - Redhat 9.0 Standart
    Linux - Fedora 2 (Redhat) (gcc 3.3.3 and gcc-3.4.1 with
                stack-smashing protector from www.trl.ibm.com/projects/security/ssp/)
    Linux - Mandrake 10.0 Official
    Linux - Ubuntu 8.04
    FreeBSD 5.2.1 (gcc 3.3.3)
    MacOS 10.4.8
    NetBSD
    CYGWIN_NT-5.1 1.5.9(0.112/4/2) 2004-03-18 23:05 i686 Cygwin
    Solaris  9 (x86) with gcc 3.3.2
    Solaris 10 (x86)

   - alpha CPU :
    Linux - Debian 3.0 (alpha) with gcc 2.95.4

   - AMD64 CPU :
    Linux - SuSE 8 ES (AMD64 Opteron) with gcc 3.2.2
    Linux - Fedora 4  (AMD64 Turion)  with gcc 4.0.1
    Linux - Ubuntu 8.10

   - Itanium CPU :
        HP-UX B.11.31 U ia64 with aCC (HP C/aC++ B3910B A.06.14 [Feb 22 2007])

   - s390x CPU :
    Linux - SUSE Enterprise Linux 10 - with gcc 4.1.2
    ( only the 32bits built works, the 64bits built does not pass the tests )

   - sparc CPU :
    Solaris 9 (sparc) with gcc 3.3.2

   - sparc CPU :
    Solaris 8 (sparc) with gcc 2.95.2


   - powerpc CPU :
    MacOS X 10.1/darwin 5.5 with gcc 932.1 (gcc 2.95.2)
    MacOS X 10.3 with XCode 1.5

   - openpower CPU :
        Linux openpower-linux1 2.6.5-7.97-pseries64 (ppc64) with gcc 3.3.3

   - IBM :
    AIX 5.3 with gcc 4.1.0

   - Tru64 :
    OSF 5.1 with gcc 3.4.2

   - PA-RISC :
    HP-UX 11.11 with aCC (HP ANSI C++ B3910B A.03.73)
    HP-UX 11.11 with gcc 3.4.2 (7za and 7zr : OK, 7z : failed because of a buggy gcc for HP-UX,
                                    plugins don't work because C++ constructors are not called !)


BUILD with cmake
----------------
  You can only build 7za.

  cd CPP/7zip/CMAKE

  ./generate.sh  (or ./generate_xcode.sh for MacOSX)
  
  In CPP/7zip, you now have directories like :
   - P7ZIP.codeblocks/
   - P7ZIP.EclipseCDT4/
   - P7ZIP.Unix/
   (or P7ZIP.Xcode/  for MacOSX)
   
   
  For the target "Unix Makefiles" :
	- cd CPP/7zip/P7ZIP.Unix
	- make
	
  Remark : these targets use only C/C++ code. To build targets with ASM code you must read the previous paragraph.	

  For others target, use the IDE (CodeBlocks, KDevelop3, Eclipse CDT4)

INSTALL :
---------

  method 1
  --------
  - edit install.sh to change DEST_HOME
  - ./install.sh : to install
  Remark : you must be "root" to install 7za in the directory "/usr/local"

  method 2
  --------
  - 7za is a stand-alone program, you can put this program where you want.
  example :  cp bin/7za /usr/local/bin/7za

  - 7z needs its plugins. You must copy the file 7z, 7z.so
    and the directory Codecs in the same destination directory.

  - if you want to be able to create SFX archive, copy bin/7zCon.sfx
  to the directory where 7za or 7z can be found.


USAGE:
------
  Remark : you can replace 7za with 7z.

  7za t archive.7z  : tests files in the archive archive.7z
  7za l archive.7z  : lists all files from archive archive.7z
  7za x archive.7z  : extracts all files from archive archive.7z
                   to the current directory

  7za a archive.7z file1 fileN : add files to the archive archive.7z

  7za a archive.7z  dir1 :  add all files and subdirectories from directory "dir1" to archive archive.7z

  CAUTION : do not use the flag "-r" unless you know what you are doing ...

  7za a -sfx archive.x dir1 : add all files from directory "dir1" to SFX archive archive.x

  You can also add or extract files to/from .zip or .tar archives.
  You can also compress/uncompress .gz or .bz2 files.

  7za a -tzip   archive.zip file file2 file3
  7za a -ttar   archive.tar file
  7za a -tgzip  file.gz     file
  7za a -tbzip2 archive.bz2 file

  You can use password for archives:
  7za a -pmy_password archive.7z  dir1

  For .7z archive only, you can enable archive header encryption with "-mhe"
  7za a -mhe=on -pmy_password archive.7z  dir1

  For more, see the documentation of 7za.exe in DOC/MANUAL directory
  7za a -t7z  -m0=lzma -mx=9 -mfb=64 -md=32m -ms=on archive.7z  dir1
      adds all files from directory "dir1" to archive archive.7z
      using "ultra settings".
      -t7z     : 7z archive
      -m0=lzma : lzma method
      -mx=9    : level of compression = 9 (Ultra)
      -mfb=64  : number of fast bytes for LZMA = 64
      -md=32m  : dictionary size = 32 megabytes
      -ms=on   : solid archive = on

  To use the new codec "lzma2"
  7za a -m0=lzma2 archive.7z directory


BENCHMARKS :
------------
  7za b : bench LZMA codec
  7z b -mm=crc : make internal crc check.
  7za b "-mm=*" : test all methods



hugetlbfs (large pages) :
-------------------------

  Using huges pages in Linux requires some preparations.
  First, make sure your running kernel has support for hugetlbfs compiled in:

  ~$ grep hugetlbfs /proc/filesystems
  nodev       hugetlbfs

  You can view your current huge page configuration like this:

  ~$ grep Huge /proc/meminfo
  HugePages_Total:       0
  HugePages_Free:        0
  HugePages_Rsvd:        0
  HugePages_Surp:        0
  Hugepagesize:       2048 kB

  In this case the size of a huge page is 2 MiB.
  So, if you have 2 GiB of RAM and want to reserve 512 MiB for huge pages,
  you would need 256 pages.

  Do the following as root:

  ~# echo 296 >/proc/sys/vm/nr_hugepages
  ~# grep Huge /proc/meminfo
  HugePages_Total:     296
  HugePages_Free:      296
  HugePages_Rsvd:        0
  HugePages_Surp:        0
  Hugepagesize:       2048 kB

  Finally, make access from user space possible:

  ~# mkdir /hugepages
  ~# mount -t hugetlbfs -o rw,nosuid,nodev,noexec,noatime none /hugepages
  ~# chmod 1777 /hugepages

  Option 1 : Now huge pages are configured.
           In your shell, set the environment variable HUGETLB_PATH to the mount point:
           ~$ export HUGETLB_PATH=/hugepages
  or

  Option 2 : Let p7zip find out the first directory with the type "hugetlbfs"


  To enable huge page use in p7zip, pass the '-slp' switch to it.
  Ex: 7za -slp a t.7z dir



  If you want this to be a permanent setup :
  - mkdir /hugepages

  - add an entry to your /etc/fstab so that when you reboot your computer,
    your hugetlbfs will mount automatically :
    hugetlbfs       /hugepages  hugetlbfs     rw,nosuid,nodev,noexec,noatime   0 0

  - add this in your /etc/rc.local :
    chmod 1777 /hugepages
    echo 296 >  /proc/sys/vm/nr_hugepages

  optional : export HUGETLB_PATH=/hugepages

  To enable huge page use in p7zip, pass the '-slp' switch to it : 7za -slp b


  Remark : with current CPU : "7za -slp b" is not faster than "7za b".
           If you want to see some speedup, compress at least 300MB
           Ex : 7za -slp -mx=9 -a big_file.tar.7z  big_file.tar


LICENSE :
---------
  please read DOC/License.txt.

LIMITATIONS from 7-zip :
------------------------

  - does not support uid/gid for the .tar format (so, use tar on Unix)

LIMITATIONS for Unix version only :
-----------------------------------

  - can handle properly UNICODE filenames only if the environment is UTF-8.
  Example : export LANG=en_US.UTF-8 ; unset LC_CTYPE
  Remark  : see the possible values for LANG in the directory
            - /usr/lib/locale   : Fedora 2 / Solaris 8
            - /usr/share/locale : OpenBSD / Debian / FreeBSD / MacOS X / Mandrake 10.1 / NetBSD

    On Linux, to display the available locales : locale -a

  p7zip relies on LC_CTYPE and then on LANG to convert name to/from UNICODE.
  the command "locale" should display these environment variables.
  you can also do "echo $LC_CTYPE" and "echo $LANG".

  If you do not plan to export your archive, you can use the flag "-no-utf16".
  Usage :
    7za a -no-utf16 archive.7z dirOrFile1 ... dirOrFileN
    7za t -no-utf16 archive.7z
    7za l -no-utf16 archive.7z
    7za x -no-utf16 archive.7z

  - ignores Windows file access permissions (files are created with default permissions)

  see also TODO file.

DEVELOPPER CORNER:
------------------

  - static analysis with cppcheck
    make cppcheck

  - static analysis with clang
    scan-build make all4 2>&1 | tee -i scan-build.out

  - dynamic analysis : valgrind
    cp makefile.linux_valgrind makefile.machine
    make test
    
  - dynamic analysis : gcc sanitize
    cp makefile.linux_amd64_sanitizer makefile.machine
    make test

  - WaitForMultipleObject has no equivalence on Unix.
  - Events don't exist.
  - sizeof(wchar_t) = 4 with GCC (2 with MS VC++)
  - "FIXME" in source code indicates that you should add codes to better handle all cases.

  - "FIXED" in source code indicates that the original code has been changed to work in Unix environment.

  see also TODO file.


ChangeLog
Version 17.05
=============

- add UTF-8 support for Client7z
- fix issue 130

Version 17.04
=============

- add lzip decompress
- update zstd 1.4.9

Version 17.03
=============

  - add zstd to zip
  - add hash algorithm
    - XXHASH32
    - XXHASH64
    - MD2
    - MD4
    - MD5
    - sha384
    - sha512
  - add Lzfse to dmg 
  - add lz5 to 7z
  - add lizard to 7z
  - update lz4 to v1.9.3
  - add brotli to 7z
  - update cmake build
  - fix xz crc64 error
  - use system local to select OEM code
  - add rpm install
  - fix lzma2 and flzma22 call the same algorithm
  - add 7zr build
  - fix tar format link file compress and decompress

Version 17.02
=============

  - p7zip 17.02 is more like 7zip 17.01(only 7za and 7z),The difference from 7zip 17.01 
    and older version p7zip is the following description
  - Supports Fast lzma2 1.0.1 compression method
  - Update Zstd method to 1.4.5
  - Add zstd method parameters in 7z format
    ZSTD parameters NEW name:
         strategy -> strat
         fast -> fast
         long -> long
         WindowLog -> wlog
         HashLog -> hlog
         ChainLog -> clog
         SearchLog -> slog
         MinMatch -> slen
         TargetLen -> tlen
         OverlapLog -> ovlog
         LdmHashLog -> ldmhlog
         LdmSearchLength -> ldmslen
         LdmBucketSizeLog -> ldmblog
         LdmHashRateLog -> ldmhevery
  - Fix symlink files contained inside tar and squashfs as regular file
  - Add lz4 and Zstd decompress method to squashfs

Version 17.01
=============

  - Fix BUG CVE-2018-10115
  - Fix BUG CVE-2018-5996
  - Fix BUG CVE-2017-17969
  - Fix BUG CVE-2016-9296

Version 17.00
=============
  
  - Supports Zstd 1.4.4 compression method
  - Supports Lz4 1.9.2 compression method

Version 16.02
=============

  - From Windows version of 7-Zip 16.02:
    - The BUG in 16.00 - 16.01 was fixed: 7-Zip mistakenly reported the warning 
      "There are some data after the end of the payload data" for split archives.

Version 16.01 (never published)
=============

  - From Windows version of 7-Zip 16.01:
    - The bugs in SWM (WIM), EXE (PE) and CHM code were fixed.
    - there are some internal changes in source code for better compatibility with VS2015 C++ compiler.

Version 16.00 (never published)
=============

  - better support for OpenBsd (CPP/Windows/System.cpp), thanks Josh (https://sourceforge.net/p/p7zip/discussion/383043/thread/ee32dcd8/?limit=25#c322)

  - From Windows version of 7-Zip 16.00:
    - 7-Zip now can extract multivolume ZIP archives (z01, z02, ... , zip).
    - Some fixed bugs:
      - bzip2 decoder -mmt2 reported E_FAIL (for k_My_HRESULT_WritingWasCut case), if we extract partial file.
      - 7z solid update (hang in break)
      - sha1 worked incorrectly for call after call with ((size & 3) != 0)
      - 7z update bcj bugs were fixed.
      - split (aaa.001) fixed
      - iso loop fix
      - rar4 multivol -stdin kpidSize
      - drag and drop 1<2.txt
      - memory access violation fix



Version 15.14.1
===============

  - patch #32 Compiling in OS X fails with p7zip_15.14 


Version 15.14
=============

  - bug #147 Directory traversal through symlinks
    ( https://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2015-1038 )

  - patch #26 : Patch to correctly set permissions for docs that permits non-sudo deletion of docs 

  - patch #27 : Update Mac OS X LLVM 64bits Makefile 

  - patch #28 : p7zip version 15.09 fails to build on s390 

  - patch #29 : please fix incorrect fsf address 

  - patch #31 : p7zip 15.09 fails to build on mips 

  - bug :  "7za a -p ..." was fixed

  - bug #172 : Impossible to create archive with some unicode symbols in filename 

  - From Windows version of 7-Zip 15.13:
    - The BUG in 15.13 in CAB code was fixed.
    
  - From Windows version of 7-Zip 15.12:
    - Some bugs were fixed.
    - New localization: Yoruba.

  - test_emul becomes test_lib  (tests libraries)

  - cross building added :
		makefile.linux_cross_mipsel

  - support for android 5+ (-fPIE -pie)

Version 15.12 (never published)
=============
  - From Windows version of 7-Zip 15.12 :
    - "There are no errors" string after "Test" operation inside archive.
    - The bugs in LZMA SDK were fixed (but these bugs are not related directly to 7-Zip's code).


  - From Windows version of 7-Zip 15.11 :
        - Some bugs were fixed.
        - 7-Zip 15.10 showed incorrect error message about missing volume for multivolume RAR archives.

   - ..../LZHAM added
   


Version 15.10 beta
==================

  - bug #147 Directory traversal through symlinks 
    Bug-Debian: https://bugs.debian.org/774660
    Delay creation of symlinks to prevent arbitrary file writes (CVE-2015-1038)
    Fix given by Ben Hutchings

  - From Windows version of 7-zip 15.10 beta:
    - The BUG in 9.21 - 15.09 was fixed:
      7-Zip could ignore some parameters, specified for archive creation operation
	  for gzip and bzip2 formats in "Add to Archive" window and in command line 
	  version (-m switch).
    - Some bugs were fixed.
      - extracting from solid wim archives worked incorrectly in some cases,
    - Also there are some minor changes.
      - 7-Zip can show the name of missing volume for multivolume RAR and VMDK archives.
      - Some internal changes with 7-Zip Benchmark.

Version 15.09 beta
==================

  - From Windows version of 7-zip 15.09 beta:
    - 7-Zip now can extract ext2 and multivolume VMDK images.
    - Some bugs were fixed.


Version 15.08 beta
==================

  - From Windows version of 7-zip 15.08 beta:

    - 7-Zip now can extract ext3 and ext4 (Linux file system) images.
    - Some bugs were fixed.


Version 15.07 beta
==================

  - "bin/Codecs/Rar29.so" renamed to "bin/Codecs/Rar.so"
  
  - support for cygwin 64 bits
  
  - support for cygwin 64 bits with asm
  
  - cygwin : fix in GetRamSize()
  
  - cross building added :
		makefile.linux_cross_aarch64
		makefile.linux_cross_arm
		makefile.linux_cross_ppc

C/Ppmd7.h
/* Ppmd7.h -- PPMdH compression codec
2016-05-21 : Igor Pavlov : Public domain
This code is based on PPMd var.H (2001): Dmitry Shkarin : Public domain */

/* This code supports virtual RangeDecoder and includes the implementation
of RangeCoder from 7z, instead of RangeCoder from original PPMd var.H.
If you need the compatibility with original PPMd var.H, you can use external RangeDecoder */

#ifndef __PPMD7_H
#define __PPMD7_H

#include "Ppmd.h"

EXTERN_C_BEGIN

#define PPMD7_MIN_ORDER 2
#define PPMD7_MAX_ORDER 64

#define PPMD7_MIN_MEM_SIZE (1 << 11)
#define PPMD7_MAX_MEM_SIZE (0xFFFFFFFF - 12 * 3)

struct CPpmd7_Context_;

typedef
  #ifdef PPMD_32BIT
    struct CPpmd7_Context_ *
  #else
    UInt32
  #endif
  CPpmd7_Context_Ref;

typedef struct CPpmd7_Context_
{
  UInt16 NumStats;
  UInt16 SummFreq;
  CPpmd_State_Ref Stats;
  CPpmd7_Context_Ref Suffix;
} CPpmd7_Context;

#define Ppmd7Context_OneState(p) ((CPpmd_State *)&(p)->SummFreq)

typedef struct
{
  CPpmd7_Context *MinContext, *MaxContext;
  CPpmd_State *FoundState;
  unsigned OrderFall, InitEsc, PrevSuccess, MaxOrder, HiBitsFlag;
  Int32 RunLength, InitRL; /* must be 32-bit at least */

  UInt32 Size;
  UInt32 GlueCount;
  Byte *Base, *LoUnit, *HiUnit, *Text, *UnitsStart;
  UInt32 AlignOffset;

  Byte Indx2Units[PPMD_NUM_INDEXES];
  Byte Units2Indx[128];
  CPpmd_Void_Ref FreeList[PPMD_NUM_INDEXES];
  Byte NS2Indx[256], NS2BSIndx[256], HB2Flag[256];
  CPpmd_See DummySee, See[25][16];
  UInt16 BinSumm[128][64];
} CPpmd7;

void Ppmd7_Construct(CPpmd7 *p);
Bool Ppmd7_Alloc(CPpmd7 *p, UInt32 size, ISzAllocPtr alloc);
void Ppmd7_Free(CPpmd7 *p, ISzAllocPtr alloc);
void Ppmd7_Init(CPpmd7 *p, unsigned maxOrder);
#define Ppmd7_WasAllocated(p) ((p)->Base != NULL)


/* ---------- Internal Functions ---------- */

extern const Byte PPMD7_kExpEscape[16];

#ifdef PPMD_32BIT
  #define Ppmd7_GetPtr(p, ptr) (ptr)
  #define Ppmd7_GetContext(p, ptr) (ptr)
  #define Ppmd7_GetStats(p, ctx) ((ctx)->Stats)
#else
  #define Ppmd7_GetPtr(p, offs) ((void *)((p)->Base + (offs)))
  #define Ppmd7_GetContext(p, offs) ((CPpmd7_Context *)Ppmd7_GetPtr((p), (offs)))
  #define Ppmd7_GetStats(p, ctx) ((CPpmd_State *)Ppmd7_GetPtr((p), ((ctx)->Stats)))
#endif

void Ppmd7_Update1(CPpmd7 *p);
void Ppmd7_Update1_0(CPpmd7 *p);
void Ppmd7_Update2(CPpmd7 *p);
void Ppmd7_UpdateBin(CPpmd7 *p);

#define Ppmd7_GetBinSumm(p) \
    &p->BinSumm[(unsigned)Ppmd7Context_OneState(p->MinContext)->Freq - 1][p->PrevSuccess + \
    p->NS2BSIndx[Ppmd7_GetContext(p, p->MinContext->Suffix)->NumStats - 1] + \
    (p->HiBitsFlag = p->HB2Flag[p->FoundState->Symbol]) + \
    2 * p->HB2Flag[(unsigned)Ppmd7Context_OneState(p->MinContext)->Symbol] + \
    ((p->RunLength >> 26) & 0x20)]

CPpmd_See *Ppmd7_MakeEscFreq(CPpmd7 *p, unsigned numMasked, UInt32 *scale);


/* ---------- Decode ---------- */

typedef struct
{
  UInt32 (*GetThreshold)(void *p, UInt32 total);
  void (*Decode)(void *p, UInt32 start, UInt32 size);
  UInt32 (*DecodeBit)(void *p, UInt32 size0);
} IPpmd7_RangeDec;

typedef struct
{
  IPpmd7_RangeDec p;
  UInt32 Range;
  UInt32 Code;
  IByteIn *Stream;
} CPpmd7z_RangeDec;

void Ppmd7z_RangeDec_CreateVTable(CPpmd7z_RangeDec *p);
Bool Ppmd7z_RangeDec_Init(CPpmd7z_RangeDec *p);
#define Ppmd7z_RangeDec_IsFinishedOK(p) ((p)->Code == 0)

int Ppmd7_DecodeSymbol(CPpmd7 *p, IPpmd7_RangeDec *rc);


/* ---------- Encode ---------- */

typedef struct
{
  UInt64 Low;
  UInt32 Range;
  Byte Cache;
  UInt64 CacheSize;
  IByteOut *Stream;
} CPpmd7z_RangeEnc;

void Ppmd7z_RangeEnc_Init(CPpmd7z_RangeEnc *p);
void Ppmd7z_RangeEnc_FlushData(CPpmd7z_RangeEnc *p);

void Ppmd7_EncodeSymbol(CPpmd7 *p, CPpmd7z_RangeEnc *rc, int symbol);

EXTERN_C_END
 
#endif
